repair_shards.c

```diff
@@ -209,3 +204,67 @@ RecreateTableDDLCommandList(Oid relationId, int64 shardId)
 
 	return extendedRecreateCommandList;
 }
+
+
+/*
+ * CopyDataFromFinalizedPlacement connects to an unhealthy placement and
+ * directs it to copy the specified shard from a certain healthy placement.
+ * This function assumes that the unhealthy placement already has a schema
+ * in place to receive rows from the healthy placement. This function returns
+ * a boolean indicating success or failure.
+ */
+static bool
+CopyDataFromFinalizedPlacement(ShardPlacement *placementToRepair,
+							   ShardPlacement *healthyPlacement,
+							   Oid distributedTableId, int64 shardId)
+{
+	char *relationName = get_rel_name(distributedTableId);
+	char relationKind = get_rel_relkind(distributedTableId);
+	const char *shardName = NULL;
+	char *nodeName = placementToRepair->nodeName;
+	int32 nodePort = placementToRepair->nodePort;
+	StringInfo copyRelationQuery = makeStringInfo();
+
+	PGconn *connection = NULL;
+	PGresult *result = NULL;
+	char *copySuccessfulString = NULL;
+	bool responseParsedSuccessfully = false;
+
+	if (relationKind == RELKIND_FOREIGN_TABLE)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot repair shard"),
+						errdetail("Repairing shards backed by foreign tables is "
+								  "currently unsupported.")));
+	}
+
+	AppendShardIdToName(&relationName, shardId);
+	shardName = quote_identifier(relationName);
+
```

Comment:
Two things:

  * I'm using a `SELECT` to call a function here, so `ExecuteRemoteCommand` doesn't work (it requires commands that return an OK status, not tuples)
  * Even if we could reduce the lines of code by calling that, this function does enough that inlining it would be very distracting. I think it needs to stay as-is.